src/components/progress-report/workflow/handlers/progress-report-workflow-notification.handler.ts

```diff
@@ -0,0 +1,231 @@
+  email?: string | undefined;
+  role?: Role[] | undefined;
+}
+type Recipient = Omit<Member, 'role'>;
+
+@EventsHandler(WorkflowUpdatedEvent)
+export class ProgressReportWorkflowNotificationHandler
+  implements IEventHandler<WorkflowUpdatedEvent>
+{
+  constructor(
+    private readonly auth: AuthenticationService,
+    private readonly repo: ProgressReportWorkflowRepository,
+    private readonly configService: ConfigService,
+    private readonly userService: UserService,
+    private readonly projectService: ProjectService,
+    private readonly languageService: LanguageService,
+    private readonly reportService: PeriodicReportService,
+    private readonly emailService: EmailService,
+    private readonly workflowService: ProgressReportWorkflowService,
+    @Logger('ProgressReportWorkflowNotificationHandler:report-status-changed')
+    private readonly logger: ILogger,
+  ) {}
+
+  async handle({
+    reportId,
+    previousStatus,
+    next,
+    session,
+    workflowEvent,
+  }: WorkflowUpdatedEvent) {
+    if (!this.configService.sendProgressReportNotifications) {
+      return;
+    }
+
+    const membersWithValidRoles = await this.getTeamMembersWithValidRoles(
+      reportId,
+      next,
+    );
+
+    const envEmailList = this.getEnvList(next) as [];
+    const usersInEnvList: Recipient[] = await Promise.all(
+      envEmailList.map(async (item) => {
+        const user = await this.repo.getUserInfoByEmail(item);
+        if (user.length > 0) {
+          const [{ id, email }] = user;
+          return { id: id, email: email };
+        } else {
+          return { id: undefined, email: item };
+        }
+      }),
+    );
+
+    const recipients = uniqBy(
+      [...usersInEnvList, ...membersWithValidRoles],
+      'email',
+    );
+    const notifications: EmailReportStatusNotification[] = await Promise.all(
+      recipients.map((recipient) =>
+        this.prepareNotificationObject(
+          session,
+          reportId,
+          previousStatus,
+          recipient,
+          workflowEvent,
+        ),
+      ),
+    );
+
+    this.logger.info('Notifying', {
+      emails: notifications.map((r) => r.recipient.email.value),
+      projectId: notifications[0]?.project.id ?? undefined,
+      languageId: notifications[0]?.language.id ?? undefined,
+      reportId: reportId,
+      newStatusVal: notifications[0]?.newStatusVal ?? undefined,
+      previousStatusVal: notifications[0]?.previousStatusVal ?? undefined,
+    });
+
+    for (const notification of notifications) {
+      if (notification.recipient.email.value) {
+        await this.emailService.send(
+          notification.recipient.email.value,
+          ProgressReportStatusChanged,
+          notification,
+        );
+      }
+    }
+  }
+
+  private async prepareNotificationObject(
+    session: Session,
+    reportId: ID,
+    previousStatus: Status,
+    receiver: Recipient,
+    unsecuredEvent: UnsecuredDto<ProgressReportWorkflowEvent>,
```

Comment:
Just pass the event.
```suggestion
    receiver: Recipient,
    event: WorkflowUpdatedEvent,
```

Reply:
I eliminated the session because I can retrieve from the workflowEvent, but I need the previousStatus since its not located anywhere else, and I need the reportId (and just tweaked the code to pass in the projectId and languageId per your other feedback) so I can generate the pertinent objects using the recipientSession.